PROTOCOL.md

```diff
@@ -0,0 +1,99 @@
+
+The structure of each client side message is:
+```typescript
+export interface OperationMessage {
+  payload?: any;
+  id?: string;
+  type: string;
+}
+```
+
+### Client -> Server
+
+#### GQL_CONNECTION_INIT
+Client sends this message after plain websocket connection to start the communication with the server
+
+This message also triggers the `onConnect` callback on the server.
+
+- `payload: Object` : optional parameters that the client specifies in `connectionParams`
+
+#### GQL_CONNECTION_TERMINATE
+Client sends this message to terminate the connection.
+
+#### GQL_START
+Client sends this message to execute GraphQL operation
+- `id: string` : The id of the GraphQL operation to start
+- `payload: Object`:
+    * `query: string` : GraphQL operation as string or parsed GraphQL document node
+    * `variables?: Object` : Object with GraphQL variables
+    * `operationName?: string` : GraphQL operation name
+    * `context?: any` : Execution context for the operation
+    
+#### GQL_STOP
+Client sends this message in order to stop a running GraphQL operation execution (for example: unsubscribe)
+- `id: string` : operation id
+    
+### Server -> Client
+
+#### GQL_CONNECTION_ERROR
+The server sends this message if `onConnect` callback returns `false` or throws an exception, and after sending this message, the server disconnects the client.
+
+It server also respond with this message in case of a parsing errors of the message (which does not disconnect the client, just ignore the message).
+
+- `payload: Object`: the server side error
+
+#### GQL_CONNECTION_ACK
+The server sends this message if `onConnect` callback returns any other value then `false`.
+
+#### GQL_DATA
+The server sends this message to transfter the GraphQL execution result from the server to the client.
+- `id: string` : ID of the operation that was successfully set up
+- `payload: Object` : 
+    * `data: any`: Execution result
+    * `errors?: Error[]` : Array of resolvers errors
+
+#### GQL_ERROR
+Server sends this message upon failing to **register** or **execute** a GraphQL operation (resolver errors are part of `GQL_DATA` message, and will be added as `errors` array)
+- `payload: { errors: Array<Object> }` : payload with an array of errors attributed to the subscription failing on the server
+- `id: string` : operation ID of the operation that failed on the server
+
+#### GQL_COMPLETE
+Server sends this message to indicate that a GraphQL operation is done, and no more data will arrive (for queries and subscription, it called immediatly, and for subscriptions it called when unsubscribing)
+- `id: string` : operation ID of the operation that completed
+
+#### GQL_CONNECTION_KEEP_ALIVE
+Server message sent periodically to keep the client connection alive.
+
+The client starts to considerer the keep alive message only upon the first received keep alive message from the server.
+
+### Messages Flow
```

Comment:
Would be great to add here (Maybe in the future and not as part of this PR) a visual representation of the states.